src/libraries/System.Private.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/AnyRef.cs

```diff
@@ -21,9 +23,39 @@ internal AnyRef(IntPtr jsHandle, bool ownsHandle) : base(ownsHandle)
         {
             SetHandle(jsHandle);
             AnyRefHandle = GCHandle.Alloc(this, ownsHandle ? GCHandleType.Weak : GCHandleType.Normal);
+            InFlight = null;
+            InFlightCounter = 0;
         }
         internal int Int32Handle => (int)(IntPtr)AnyRefHandle;
 
+        internal void AddInFlight()
+        {
+            lock (this)
+            {
+                InFlightCounter++;
+                if (InFlightCounter == 1 && InFlight == null)
+                {
+                    InFlight = GCHandle.Alloc(this, GCHandleType.Normal);
+                }
+            }
+        }
+
+        internal void ReleaseInFlight()
+        {
+            lock (this)
+            {
+                Debug.Assert(InFlightCounter != 0);
+
+                InFlightCounter--;
+                if (InFlightCounter == 0 && InFlight.HasValue)
```

Comment:
Same here, the defensive .HasValue check would actually prevent us from detecting corrupt state since we're not throwing if it's wrong

Reply:
changed it into assert